docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,446 @@
+
+#### Considerations about the CRD design
+
+The scope of the cluster values accessible using Go templating must be defined along with use cases. An option would be to expose the Builtin struct type used in ClusterClass or construct a similar type. Additionally, while resources like the control plane and cluster have a 1:1 correlation with each selected cluster, resources like MachineDeployments or Machines could have any number and can’t be selected in a similar way. As a result, additional work is needed to decide if there’s a use case for including information in those types.
+
+It would also be useful to add `clusterv1.Conditions` to the HelmChartProxy and HelmReleaseProxy status fields to surface additional information about the status of the Helm release in a way that is consistent with the design of existing Cluster API CRDs.
+
+### Example: Installing Calico CNI
+
+Developers and Cluster operators often want to install a CNI on their workload clusters. HelmChartProxy controller can be used to install the Calico CNI on each selected workload cluster. In this example, the value configuration can be leveraged to initialize the `installation.calicoNetwork.ipPools` array by creating an element for every pod CIDR in the Cluster.
+
+```yaml
+apiVersion: addons.cluster.x-k8s.io/v1beta1
+kind: HelmChartProxy
+metadata:
+  name: calico-cni
+spec:
+  clusterSelector:
+    key: calicoCNI
+    value: enabled
+  releaseName: calico
+  repoURL: https://projectcalico.docs.tigera.io/charts
+  chartName: tigera-operator
+  values: |
+    installation:
+      cni:
+        type: Calico
+        ipam:
+          type: HostLocal
+      calicoNetwork:
+        bgp: Disabled
+        mtu: 1350
+        ipPools:{{range $i, $cidr := .Cluster.Spec.ClusterNetwork.Pods.CIDRBlocks }}
+        - cidr: {{ $cidr }}
+          encapsulation: None
+          natOutgoing: Enabled
+          nodeSelector: all(){{end}}
+```
+
+## Cluster API enhancements
+
+Despite the fact that most of the heavy lifting for add-on management is delegated to external ClusterAddonProvider components, over time we would like to provide Cluster API’s users an integrated and seamless user experience for add-on orchestration.
+
+### clusterctl
+
+ClusterAddonProvider is a new component to be installed in a Management Cluster, and in order to make this operation simpler, we can extend clusterctl in order to manage ClusterAddonProvider implementations in clusterctl init, clusterctl upgrade, etc.
+
+Note: clusterctl assumes all the providers to be linked to a CAPI contract; while it is not clear at this stage if the same applies to ClusterAddonProvider. This should be further investigated.
+
+## Alternatives
+
+### Why not CRS
+
+Here are reasons why ClusterResourceSets are not a good option for cluster-api-idiomatic addons:
```

Comment:
I think we are missing some of the most important ones:
* In order to improve the add-on management capabilities of CRS, we would need to re-invent the wheel, doing what a few other tools in the space are already doing.
* Users are already using add-ons management tools and they want to use the same ones for cluster add-ons.
* CRS design presents severe limitations around deleting resources and dynamic versioning based on the kubernetes version.

not blocking: some of the reasons listed here are a bit sketchy, like `require a ton of effort to do` or `require significant refactor work that may impact existing CRS users`. One could argue that implementing a `ClusterAddonProvider` is as complex or more than such refactor. No need to remove them (I 💯 agree ClusterAddonProvider is a better solution to the problem), but from a "why didn't we do this" debate perspective, some of these sound a bit forced.